File: mirea/cources/beginner/connected-components-3/a.py
# структура для системы непересекающихся множеств
class DSU:
    
    def __init__(self, n):
        self.sz = [1] * n;
        self.par = [int(x) for x in range(n)];
    
    def getPar(self, u):
        if u == self.par[u]:
            return u;
        p = self.getPar(self.par[u]);
        self.par[u] = p;
        return p;
    
    def connect(self, u, v):
        u, v = self.getPar(u), self.getPar(v);
        if u != v:
            if self.sz[u] < self.sz[v]:
                u, v = v, u; # swap
            self.sz[u] += self.sz[v];
            self.par[v] = u;

# оставить только уникальные элементы в сортированном списке
def unique(s):
    i, p, n = 0, 0, len(s);
    while (i < n):
        j = i;
        while (j < n and s[i] == s[j]):
            j += 1;
        s[p] = s[i];
        i = j;
        p += 1;
    while (n > p):
        s.pop();
        n -= 1;

# пробегаемся по всем подстрокам длины k и проводим рёбра
# между позициями символов, которые должны быть равны
# это означает, что на позициях, соединённых рёбрами, должны стоять
# одинаковые символы. Компонента связности содержит все позиции
# символов, которые должны быть попарно-равны между собой
n,m,k = (int(x) for x in input().split())
dsu = DSU(n);
for start in range(0,n-k+1):
    for i in range(0,k//2):
        dsu.connect(start+i, start+k-1-i);
# теперь считаем количество различных компонент
parents = [];
for i in range(0, n):
    parents.append(dsu.getPar(i));
parents.sort();
unique(parents);
# выводим ответ: для каждой компоненты можем выбрать один из m символов
# произвольно, поэтому ответ равен m в степени количества компонент:
print(pow(m,len(parents),1000000007));

File: mirea/cources/beginner/connected-components-3/a.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

// структура для системы неперекающихся множеств
class DSU {
public:
    vector<int> sz, par;

    DSU(int n) {
        sz.resize(n, 1);
        par.resize(n);
        for (int i = 0; i < n; i++) {
            par[i] = i;
        }
    }

    int getPar(int u) {
        if (u == par[u]) return u;
        return par[u] = getPar(par[u]);
    }

    void connect(int u, int v) {
        u = getPar(u);
        v = getPar(v);
        if (u != v) {
            if (sz[u] < sz[v]) swap(u, v); // swap
            sz[u] += sz[v];
            par[v] = u;
        }
    }
};

int main() {
    int n, m, k;
    cin >> n >> m >> k;
    DSU dsu(n);

    // пробегаемся по всем подстрокам длины k и проводим рёбра
    // между позициями символов, которые должны быть равны
    for (int start = 0; start <= n - k; start++) {
        for (int i = 0; i < k / 2; i++) {
            dsu.connect(start + i, start + k - 1 - i);
        }
    }

    // теперь считаем количество различных компонент
    vector<int> parents;
    for (int i = 0; i < n; i++) {
        parents.push_back(dsu.getPar(i));
    }
    sort(parents.begin(), parents.end());
    parents.erase(unique(parents.begin(), parents.end()), parents.end());

    // выводим ответ: для каждой компоненты можем выбрать один из m символов
    // произвольно, поэтому ответ равен m в степени количества компонент:
    const int MOD = 1000000007;
    long long result = 1;
    for (int i = 0; i < parents.size(); i++) {
        result = (result * m) % MOD;
    }
    cout << result << endl;
    return 0;
}